### PR TITLE
feat: hasRequiredApprovals built-in

### LINKED ISSUE #406: hasRequiredApprovals built-in
**Is your feature request related to a problem? Please describe.**
In GitHub, we can set a specific number of required approving reviews to all pull requests in a repository.
However, it would be useful to have a way of specifying more the required approvals depending on the scenario.

**Describe the solution you'd like**
Develop a built-in `hasRequiredApprovals` that given a X number of required approving reviews and a list of specific users, it checked if X of those users had given approving reviews to the pull request.

**Additional context**
```
hasRequiredApprovals:: (Int, []String) => Boolean
```
Use Case:
```
api-version: reviewpad.com/v3.x

rules:
  - name: changes-critical-code
    kind: patch
    description: Changes to critical code
    spec: '$hasAnnotation("critical")'

workflows:
  - name: has-required-approvals
    if:
      - '$hasRequiredApprovals(2, $team("core"))'
    then:
      - '$fail("Blocking the merge")'
```

### PR DESCRIPTION
**Pull request by shay2025**
## Description

This pull request adds the new built-in `hasRequiredApprovals` that checks if a pull request has the x approvals from a given list of users.

## Related issue

Closes #406 

## Type of change

New feature (non-breaking change which adds functionality)

## How was this tested?

Ran `task build -f` and `task test -f` cmd.

## Checklist

- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works 
- [x] I have ran `task check -f` and have no issues

## Code review and merge strategy (ship/show/ask) 

 
 
- [x] Ask: this pull request requires a code review before merge

### CODE REVIEW

File: plugins/aladino/functions/hasRequiredApprovals.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	if len(requiredApprovalsFrom) < totalRequiredApprovals {
+		return nil, fmt.Errorf("hasRequiredApprovals: the number of required approvals exceeds the number of members from the given list of required approvals")
+	}
+
+	approvedBy, err := getApprovals(target)
+	if err != nil {
+		return nil, err
+	}
+
+	var requiredApprovalsCount int
```
Comment (ferreiratiago) [inline plugins/aladino/functions/hasRequiredApprovals.go:33]: I would suggest specifying the initial value and not relying on what go uses as default values.

Also, this variable should be named `totalApprovedReviews`.

File: plugins/aladino/functions/hasRequiredApprovals.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+
+	"github.com/reviewpad/reviewpad/v3/codehost/github/target"
+	"github.com/reviewpad/reviewpad/v3/handler"
+	"github.com/reviewpad/reviewpad/v3/lang/aladino"
+)
+
+func HasRequiredApprovals() *aladino.BuiltInFunction {
+	return &aladino.BuiltInFunction{
+		Type:           aladino.BuildFunctionType([]aladino.Type{aladino.BuildIntType(), aladino.BuildArrayOfType(aladino.BuildStringType())}, aladino.BuildBoolType()),
```
Comment (ferreiratiago) [inline plugins/aladino/functions/hasRequiredApprovals.go:13]: Suggestion for better reading
```suggestion
		Type: aladino.BuildFunctionType(
			[]aladino.Type{
				aladino.BuildIntType(),
				aladino.BuildArrayOfType(aladino.BuildStringType())
			},
			aladino.BuildBoolType(),
		),
```

File: plugins/aladino/functions/hasRequiredApprovals.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		Code:           hasRequiredApprovalsCode,
+		SupportedKinds: []handler.TargetEntityKind{handler.PullRequest},
+	}
+}
+
+func hasRequiredApprovalsCode(e aladino.Env, args []aladino.Value) (aladino.Value, error) {
+	target := e.GetTarget().(*target.PullRequestTarget)
```
Comment (ferreiratiago) [inline plugins/aladino/functions/hasRequiredApprovals.go:20]: Suggestion, rename `target` to `pullRequest`. Target is not clear enough.

File: plugins/aladino/functions/hasRequiredApprovals.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+		if requiredApprovalsCount == totalRequiredApprovals {
+			return aladino.BuildBoolValue(true), nil
+		}
+	}
+
+	return aladino.BuildBoolValue(false), nil
+}
+
+func getApprovals(pullRequest *target.PullRequestTarget) ([]string, error) {
```
Comment (ferreiratiago) [inline plugins/aladino/functions/hasRequiredApprovals.go:47]: Suggestion, rename the function to `getUserLoginsFromApprovals`.

File: plugins/aladino/functions/hasRequiredApprovals.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		if contains(approvedBy, requiredApproval.(*aladino.StringValue).Val) {
+			requiredApprovalsCount++
+		}
+
+		if requiredApprovalsCount == totalRequiredApprovals {
+			return aladino.BuildBoolValue(true), nil
+		}
+	}
+
+	return aladino.BuildBoolValue(false), nil
```
Comment (ferreiratiago) [inline plugins/aladino/functions/hasRequiredApprovals.go:44]: I would suggest a different approach. IMO, the code becomes more readable, although the number of operations is higher. WDYT?

```suggestion
	var requiredApprovalsCount int
	for _, requiredApproval := range requiredApprovalsFrom {
		if contains(approvedBy, requiredApproval.(*aladino.StringValue).Val) {
			requiredApprovalsCount++
		}
	}

	return aladino.BuildBoolValue(requiredApprovalsCount > requiredApprovalsCount), nil
```

**Review by ferreiratiago — DISMISSED** [review]

**Review by marcelosousa — APPROVED** [review]

**Pull request merged**